File: src/main/java/com/bytebreak/tictactoe/rule/CannotOverwritePreviousPlaysRule.java
package com.bytebreak.tictactoe.rule;

import com.bitdecay.board.Action;
import com.bitdecay.board.Rule;
import com.bytebreak.tictactoe.State;

public class CannotOverwritePreviousPlaysRule implements Rule<State> {
    @Override
    public boolean apply(State current, State next, Action action) {
        return notOverWritten(current.q, next.q) &&
                notOverWritten(current.w, next.w) &&
                notOverWritten(current.e, next.e) &&
                notOverWritten(current.a, next.a) &&
                notOverWritten(current.s, next.s) &&
                notOverWritten(current.d, next.d) &&
                notOverWritten(current.z, next.z) &&
                notOverWritten(current.x, next.x) &&
                notOverWritten(current.c, next.c);
    }

    private boolean notOverWritten(int a, int b){
        return !(a != 0 && a != b);
    }

    @Override
    public String description() {
        return null;
    }

    @Override
    public String serialize() {
        return null;
    }

    @Override
    public Rule deserialize(String s) {
        return null;
    }
}

File: src/main/scala/com/bitdecay/tictactoe/scala/rule/CannotOverwritePreviousPlaysRule.scala
package com.bitdecay.tictactoe.scala.rule

import com.bitdecay.board.{Action, GameBoardState, Rule}
import com.bitdecay.tictactoe.scala.State


object CannotOverwritePreviousPlaysRule extends Rule[State] {
  override def apply(cur: State, next: State, action: Action[_ <: GameBoardState]): Boolean = notOverWritten(cur.q, next.q) && notOverWritten(cur.w, next.w) && notOverWritten(cur.e, next.e) && notOverWritten(cur.a, next.a) && notOverWritten(cur.s, next.s) && notOverWritten(cur.d, next.d) && notOverWritten(cur.z, next.z) && notOverWritten(cur.x, next.x) && notOverWritten(cur.c, next.c)

  private def notOverWritten(a:Int, b:Int) = !(a != 0 && a != b)

  override def description(): String = ???

  override def serialize(): String = ???

  override def deserialize(s: String): Rule[_ <: GameBoardState] = ???
}
